src/metabase/db/liquibase.clj

```diff
@@ -228,14 +232,43 @@
   [liquibase]
   (.hasChangeLogLock (lock-service liquibase)))
 
+(defn- wait-until [done? ^long sleep-ms timeout-ms]
+  (let [deadline (+ (System/nanoTime) (* 1e6 timeout-ms))]
+    (loop []
+      (if (done?)
+        :done
+        (if (>= (System/nanoTime) deadline)
+          :timed-out
+          (do (Thread/sleep sleep-ms)
+              (recur)))))))
+
+(defn- locked-instances []
+  (filter holding-lock? @potentially-locked-instances))
+
+(defn wait-for-all-locks
+  "Wait up to a maximum of `timeout-seconds` for the given Liquibase instance to release the migration lock."
+  [sleep-ms timeout-ms]
+  (log/warn (locked-instances))
+  (let [done? #(empty? (locked-instances))]
+    (if (done?)
+      :none
+      (do (log/info "Waiting for migration lock(s) to be released")
+          (wait-until done? sleep-ms timeout-ms)))))
+
+(defn release-all-locks-if-needed!
+  "Release all locks held by this process."
+  []
+  (doseq [liquibase (locked-instances)]
+    (release-lock-if-needed! liquibase)))
```

Comment:
There might be a critical issue here - by using the same `liquibase` instance as the migrations, we end up sharing the connection. I'm not sure how this would play out, but it might mean the server hangs.

I think this needs to be reworked to make sure we open new connections, but not sure how to do that in a way that supports multiple databases, as the rest of the code notionally does.

Reply:
Resolved this in a fairly clumsy way - I could not find a way to clone the connection as nothing seems to have held onto the password. Its as if these objects are meant to be secure.